Dockerfile

```diff
@@ -1,25 +1,45 @@
+## BUILDER STAGE ##
 FROM golang:1.13-alpine as builder
 RUN apk --no-cache add git
-WORKDIR /
-COPY . /
+WORKDIR /app
+
+COPY go.mod .
+COPY go.sum .
+RUN go mod download
+
+COPY . /app
+
 RUN GOOS=linux CGO_ENABLED=0 GOARCH=amd64 go build -o conftest -ldflags="-w -s" cmd/main.go
 
+## TEST STAGE ##
 FROM golang:1.13-alpine as test
-COPY --from=builder /conftest /
-RUN go test /conftest/...
+WORKDIR /app
+
+COPY go.mod .
```

Comment:
would it be possible to do this in a `module` or `vendor` stage? this way we could share the cache between build and test

Reply:
hrm.. I tried doing that, but now that you bring it up I bet it was because of the `RUN go build` command. I was effectively saying `COPY --from builder` in the test layer, because both builder and test use the same artifacts, but the test layer still re-downloaded all of the modules.

Let me see if I can split out the module download piece and if it makes sense to do so.

Reply:
I don't think it would be worth doing so. It feels like added complexity and less readibility to reduce a few lines of code.

To make this work, I'd think we'd have to do `COPY go.mod/go.sum/go mod download` and then do a `COPY` from the `vendor` layer's `go/pkg/cache` folder to copy over the downloaded modules.